debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbSnapshotChangeEventSource.java

```diff
@@ -0,0 +1,599 @@
+                throw new InterruptedException("Interrupted while awaiting initial snapshot delay");
+            }
+
+            LOGGER.info("The connector will wait for {}s before proceeding", timer.remaining().getSeconds());
+            metronome.pause();
+        }
+    }
+
+    protected SnapshotContext prepare(ChangeEventSourceContext sourceContext) throws Exception {
+        return new MongoDbSnapshotContext();
+    }
+
+    protected void complete(SnapshotContext snapshotContext) {
+    }
+
+    private void snapshotReplicaSet(ChangeEventSourceContext sourceContext, SnapshotContext ctx, ReplicaSet replicaSet) throws InterruptedException {
+        MongoPrimary primaryClient = null;
+        try {
+            primaryClient = establishConnectionToPrimary(replicaSet);
+            if (primaryClient != null) {
+                createDataEvents(sourceContext, ctx, replicaSet, primaryClient);
+            }
+        }
+        finally {
+            if (primaryClient != null) {
+                primaryClient.stop();
+            }
+        }
+    }
+
+    private MongoPrimary establishConnectionToPrimary(ReplicaSet replicaSet) {
+        return connectionContext.primaryFor(replicaSet, taskContext.filters(), (desc, error) -> {
+            // propagate authorization failures
+            if (error.getMessage() != null && error.getMessage().startsWith(AUTHORIZATION_FAILURE_MESSAGE)) {
+                throw new ConnectException("Error while attempting to " + desc, error);
+            }
+            else {
+                LOGGER.error("Error while attempting to {}: ", desc, error.getMessage(), error);
+                throw new ConnectException("Error while attempting to " + desc, error);
+            }
+        });
+    }
+
+    private boolean isInitialSyncExpected(MongoPrimary primaryClient, ReplicaSetOffsetContext offsetContext) {
+        boolean performSnapshot = true;
+        if (offsetContext.hasOffset()) {
+            if (LOGGER.isInfoEnabled()) {
+                LOGGER.info("Found existing offset for replica set '{}' at {}", offsetContext.getReplicaSetName(), offsetContext.getOffset());
+            }
+            performSnapshot = false;
+            if (connectionContext.performSnapshotEvenIfNotNeeded()) {
+                LOGGER.info("Configured to perform initial sync of replica set '{}'", offsetContext.getReplicaSetName());
+                performSnapshot = true;
+            }
+            else {
+                if (offsetContext.isInitialSyncOngoing()) {
+                    // The latest snapshot was not completed, so restart it
+                    LOGGER.info("The previous initial sync was incomplete for '{}', so initiating another initial sync", offsetContext.getReplicaSetName());
+                    performSnapshot = true;
+                }
+                else {
+                    // There is no ongoing initial sync, so look to see if our last recorded offset still exists in the oplog.
```

Comment:
I believe this deserves a TODO note - right now we implement when needed snapshot by default. In future we should provide the same otpions as other connectors and this should server as a reminder where wheen_needed functionlaity lies.

Reply:
See commit 0158b56.